### ISSUE TITLE
Tuple Filter not working as in 3.0.0-M2

### ISSUE THREAD

**Issue by scalavision** [lampepfl/dotty#10877]
## Minimized code

I am sorry I can't minimize it further, as I am learning typelevel programming in general, trying to use dotty.

The code is taken from:
https://github.com/weihsiu/tlp

In this project the HMap failes to compile similarily.

This compiles in version 3.0.0-M2, from my understanding and from reading the release changes, I can't see why it doesn't compile in M3.

```Scala
// tic tac toe coordinates
// Top Center, Middle Center, Bottom Center
sealed trait Pos
case object TC extends Pos
case object MC extends Pos
case object BC extends Pos

sealed trait Player
case object O extends Player
case object X extends Player

type OtherPlayer[P <: Player] <: Player = P match
  case O.type => X.type
  case X.type => O.type

type ContainsPos[P <: Pos] = [M] =>> ContainsPosAux[P, M]

type ContainsPosAux[P <: Pos, M] <: Boolean = M match
  case Move[_, P] => true
  case _ => false

case class Move[+PL <: Player, +PS <: Pos](player: PL, pos: PS)

case class Board[P <: Player, MS <: Tuple](moves: List[Move[Player, Pos]])

val emptyBoard = Board[O.type, EmptyTuple](Nil)

def playMove[
  P <: Player,
  MS <: Tuple,
  PL <: Player,
  PS <: Pos
](
  board: Board[P, MS], move: Move[PL, PS]
)(using P =:= PL, Filter[MS, ContainsPos[PS]] =:= EmptyTuple)
    : Board[OtherPlayer[P], Move[PL, PS] *: MS] =

  Board[OtherPlayer[P], Move[PL, PS] *: MS](move :: board.moves)

val b1 = playMove(emptyBoard, Move(O, MC))
val b2 = playMove(b1, Move(X, TC))
val b3 = playMove(b2, Move(O, BC))
```

## Output

```scala
error] -- Error: /home/yoda/Projects/evolve/tlp/src/main/scala/tlp/TicTacToe2.scala:45:36 
[error] 45 |  val b2 = playMove(b1, Move(X, TC))
[error]    |                                    ^
[error]    |Cannot prove that Tuple.Filter[
[error]    |  (tlp.TicTacToe2.Move[tlp.TicTacToe2.O.type, tlp.TicTacToe2.MC.type] *: 
[error]    |    EmptyTuple.type
[error]    |  )
[error]    |, tlp.TicTacToe2.ContainsPos[tlp.TicTacToe2.TC.type]] =:= EmptyTuple.
[error] -- Error: /home/yoda/Projects/evolve/tlp/src/main/scala/tlp/TicTacToe2.scala:46:36 
[error] 46 |  val b3 = playMove(b2, Move(O, BC))
[error]    |                                    ^
[error]    |      Cannot prove that tlp.TicTacToe2.Player =:= tlp.TicTacToe2.O.type.
```

## Expectation

Should compile

**Comment by prolativ** [comment]
This looks like some regression in the area of type aliases. This works if you inline `ContainsPos` like this
```scala
sealed trait Pos
case object TC extends Pos
case object MC extends Pos
case object BC extends Pos

sealed trait Player
case object O extends Player
case object X extends Player

type OtherPlayer[P <: Player] <: Player = P match
  case O.type => X.type
  case X.type => O.type

type ContainsPos[P <: Pos] = [M] =>> ContainsPosAux[P, M]

type ContainsPosAux[P <: Pos, M] <: Boolean = M match
  case Move[_, P] => true
  case _ => false

case class Move[+PL <: Player, +PS <: Pos](player: PL, pos: PS)

case class Board[P <: Player, MS <: Tuple](moves: List[Move[Player, Pos]])

val emptyBoard = Board[O.type, EmptyTuple](Nil)

def playMove[
  P <: Player,
  MS <: Tuple,
  PL <: Player,
  PS <: Pos
](
  board: Board[P, MS], move: Move[PL, PS]
)(using P =:= PL, Tuple.Filter[MS, [M] =>> ContainsPosAux[PS, M]] =:= EmptyTuple)
    : Board[OtherPlayer[P], Move[PL, PS] *: MS] =

  Board[OtherPlayer[P], Move[PL, PS] *: MS](move :: board.moves)


val b1 = playMove(emptyBoard, Move(O, MC))
val b2 = playMove(b1, Move(X, TC))
val b3 = playMove(b2, Move(O, BC))
```

**Comment by scalavision** [comment]
Thanks a lot! I am still not able to get it working for the HMap example. I need to learn more about  `=>>` so will try to poke around a bit :-)

**Comment by prolativ** [comment]
The same workaround works for that example:

```
import scala.Tuple.{Filter, Size}

object HMaps:
  trait HMap[K, ES <: Tuple]:
    type Entry[K, V] = (K, V)
    type Not[B <: Boolean] <: Boolean = B match
      case true => false
      case _ => true
    type Is[T1, T2] <: Boolean = T2 match
      case T1 => true
      case _ => false
    type IsC[T1] = [T2] =>> Is[T1, T2]
    type IsNotC[T1] = [T2] =>> Not[Is[T1, T2]]
    type IsKey[K, E] <: Boolean = E match
      case Entry[K, _] => true
      case _ => false
    type IsKeyC[K] = [E] =>> IsKey[K, E]
    type IsNotKeyC[K] = [E] =>> Not[IsKey[K, E]]

    def apply[V](key: K)(using 1 =:= Size[Filter[ES, [T] =>> Is[Entry[key.type, V], T]]]): V
    def put[V](key: K, value: V): HMap[K, Entry[key.type, V] *: Filter[ES, [T] =>> Not[IsKey[key.type, T]]]]
    def remove(key: K)(using 1 =:= Size[Filter[ES, [T] =>> IsKey[key.type, T]]]): HMap[K, Filter[ES, [T] =>> Not[IsKey[key.type, T]]]]

  object HMap:
    def apply[K]: HMap[K, EmptyTuple] = MapHMap[K, EmptyTuple](Map.empty[K, Any])
    private class MapHMap[K, ES <: Tuple](map: Map[K, Any]) extends HMap[K, ES]:
      def apply[V](key: K)(using 1 =:= Size[Filter[ES, [T] =>> Is[Entry[key.type, V], T]]]): V =
        map(key).asInstanceOf[V]
      def put[V](key: K, value: V): HMap[K, Entry[key.type, V] *: Filter[ES, [T] =>> Not[IsKey[key.type, T]]]] =
        MapHMap[K, Entry[key.type, V] *: Filter[ES, [T] =>> Not[IsKey[key.type, T]]]](map + (key -> value))
      def remove(key: K)(using 1 =:= Size[Filter[ES, [T] =>> IsKey[key.type, T]]]): HMap[K, Filter[ES, [T] =>> Not[IsKey[key.type, T]]]] =
        MapHMap[K, Filter[ES, [T] =>> Not[IsKey[key.type, T]]]](map - key)

  import HMap._
  val hm = HMap[String].put[String]("name", "walter").put[Int]("age", 18)
  val hm1 = hm.put[Int]("age", 20).put[String]("age", "twenty")
  println(hm1[String]("name"))
  // val age: Int = hm1("age")
  println(hm1[String]("age"))
  val hm2 = hm.remove("age")
  hm2[String]("name")
  // hm2[Int]("age")
```

**Comment by scalavision** [comment]
Thanks again :-)

**Comment by odersky** [comment]
@prolativ Can you have a look? If stuck otherwise, maybe by identifying the commit that regressed?

**Comment by prolativ** [comment]
This got fixed in 3.1.1
